### ISSUE TITLE
AGIC Removes Custom Error Pages on Listeners

### ISSUE THREAD

**Issue by kevinmarx** [Azure/application-gateway-kubernetes-ingress#733]
I have custom error pages applied to listeners in App Gateway. AGIC seems to be removing these any time it performs an update to the AGW. Is there a way I can define the custom error pages as part of the ingress resource?

**To Reproduce**
Set custom error pages on listeners for 403 and 502 errors.
Perform a kubernetes deploy
Observe the listeners no longer have custom error pages applied.

**Ingress Controller details**
```
Name:           ingress-azure-1579727113-5c8f655454-zmgvb
Namespace:      default
Priority:       0
Node:           aks-stdd4sv3-29704773-vmss000000/10.250.16.97
Start Time:     Tue, 28 Jan 2020 15:40:00 -0600
Labels:         aadpodidbinding=ingress-azure-1579727113
                app=ingress-azure
                pod-template-hash=5c8f655454
                release=ingress-azure-1579727113
Annotations:    prometheus.io/port: 8123
                prometheus.io/scrape: true
Status:         Running
IP:             10.250.16.117
Controlled By:  ReplicaSet/ingress-azure-1579727113-5c8f655454
Containers:
  ingress-azure:
    Container ID:   docker://5cf43d7fb60321928a32fff0fefdcb15d06b09108e4bdb820c44208e00cccaba
    Image:          mcr.microsoft.com/azure-application-gateway/kubernetes-ingress:1.0.0
    Image ID:       docker-pullable://mcr.microsoft.com/azure-application-gateway/kubernetes-ingress@sha256:c295f99ae66443c5a392fd894620fcd1fc313b9efdec96d13f166fefb29780a9
    Port:           <none>
    Host Port:      <none>
    State:          Running
      Started:      Tue, 04 Feb 2020 15:14:16 -0600
    Last State:     Terminated
      Reason:       Error
      Exit Code:    2
      Started:      Tue, 28 Jan 2020 15:40:25 -0600
      Finished:     Tue, 04 Feb 2020 15:13:20 -0600
    Ready:          True
    Restart Count:  1
    Liveness:       http-get http://:8123/health/alive delay=15s timeout=1s period=20s #success=1 #failure=3
    Readiness:      http-get http://:8123/health/ready delay=5s timeout=1s period=10s #success=1 #failure=3
    Environment Variables from:
      ingress-azure-1579727113  ConfigMap  Optional: false
    Environment:
      AZURE_CONTEXT_LOCATION:  /etc/appgw/azure.json
      AGIC_POD_NAME:           ingress-azure-1579727113-5c8f655454-zmgvb (v1:metadata.name)
      AGIC_POD_NAMESPACE:      default (v1:metadata.namespace)
    Mounts:
      /etc/appgw/azure.json from azure (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from ingress-azure-1579727113-token-8xfvp (ro)
Conditions:
  Type              Status
  Initialized       True
  Ready             True
  ContainersReady   True
  PodScheduled      True
Volumes:
  azure:
    Type:          HostPath (bare host directory volume)
    Path:          /etc/kubernetes/azure.json
    HostPathType:  File
  ingress-azure-1579727113-token-8xfvp:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  ingress-azure-1579727113-token-8xfvp
    Optional:    false
QoS Class:       BestEffort
Node-Selectors:  <none>
Tolerations:     node.kubernetes.io/not-ready:NoExecute for 300s
                 node.kubernetes.io/unreachable:NoExecute for 300s
Events:          <none>
```

**Comment by abenck** [comment]
@draychev 
This is a good feature to have, we just discussed doing this while performing maintenance on a production application.

**Comment by Danlewis3** [comment]
We would like this too as we use custom 502 pages from App gateway

**Comment by MarcelT-NL** [comment]
I can confirm this behavior. Breaks our custom error pages....

**Comment by manojpramesh** [comment]
Same here. I though I could get away with a [application gateway level error pages](https://docs.microsoft.com/en-sg/powershell/module/az.network/add-azapplicationgatewaycustomerror?view=azps-3.7.0&viewFallbackFrom=azps-1.2.0), but it still returns default pages.

**Comment by MXClyde** [comment]
Anybody found a reasonable workaround for getting custom error pages in place for AGIC-controlled listeners?

**Comment by Danlewis3** [comment]
Do it on the waf level not listener level

**Comment by MXClyde** [comment]
manojpramesh's comment (3 Apr) suggests that won't work. I'll be giving it a try soon...

**Comment by MarcoDuran13** [comment]
I have the same trouble. I tried the same as manojpramesh, but It still not working.

**Comment by MXClyde** [comment]
I actually got it working by just configuring it on the AG global level, with a PowerShell like:

$resourceGroup = “rg”
$AppGWName = "blablabla-ag"
$customError502Url = "https://blablabla.blob.core.windows.net/errpage/maintenance.html"
$AppGw = Get-AzApplicationGateway -Name $AppGWName -ResourceGroupName $resourceGroup
$updatedGW = Add-AzApplicationGatewayCustomError -ApplicationGateway $AppGw -StatusCode HttpStatus502 -CustomErrorPageUrl $customError502Url 
Set-AzApplicationGateway -ApplicationGateway $AppGw`

**Comment by MarcelT-NL** [comment]
Yes, but that reverts as soon as AGIC updates the gateway right?

**Comment by MarcoDuran13** [comment]
Yes MarcelT-NL. I tried it, and after the ag was update, the changes reverts. I think that is something about identity who is owner of appgw. Changes must be made with this identity.

**Comment by MXClyde** [comment]
Nope, it seems to stick in my environment...

**Comment by MarcelT-NL** [comment]
Would that be different from:

`az network application-gateway update --resource-group blah-rg --name blah-appgateway --custom-error-pages HttpStatus403=http://blah/error403.html HttpStatus502=http://blah/error502.html`

This is overridden from time to time by AGIC

**Comment by MarcelT-NL** [comment]
Trying your solution now. Any idea how to test the 502 quickly? My custom WAF seems to allow almost anything...

**Comment by MXClyde** [comment]
Scaling one of your deployments to zero pods...

**Comment by MarcelT-NL** [comment]
> Scaling one of your deployments to zero pods...

The PS command seems to work, while the AZ command reverts. I will wait a couple of days to see if it sticks!

**Comment by mhabegger** [comment]
Any news on this topic ?

**Comment by renancnetto** [comment]
Any news??

**Comment by dymitr-soltysiak-wttech** [comment]
Any updates on that opened PR? Any plans on releasing it?

**Comment by KlemenDanfoss** [comment]
How to push this forward? We also need this functionality...

**Comment by NicholasMcGrath** [comment]
Unfortunately in ps but sets a custom url at the gateway level. 

$resourceGroup = "your rg"
$AppGWName = "your gw"
$customError502Url = "your url"
$customError403Url = "your url"
$AppGw = Get-AzApplicationGateway -Name $AppGWName -ResourceGroupName $resourceGroup
$updatedGW = Add-AzApplicationGatewayCustomError -ApplicationGateway $AppGw -StatusCode HttpStatus502 -CustomErrorPageUrl $customError502Url
$updatedGW = Add-AzApplicationGatewayCustomError -ApplicationGateway $AppGw -StatusCode HttpStatus403 -CustomErrorPageUrl $customError403Url
Set-AzApplicationGateway -ApplicationGateway $AppGw

**Comment by spniak** [comment]
According to [MS doc on custom error setting](https://learn.microsoft.com/en-us/azure/application-gateway/custom-error#configuration-levels) the above script should do the work since it is a global level setting. I use AGIC and so far seems like  the solution. There are no custom error pages configured on listeners levels.

Currently to use the AZ Module or this I had to install it with verbose param. Here are my steps:
Install-Module -Name Az -Repository PSGallery -Force -Verbose
Connect-AzAccount -tenant <tenant_ID> -Subscription <subscription_ID>
$resourceGroup = "<rg_name>"
$AppGWName = "<agw_name>"
$customError502Url = "<url>"
$AppGw = Get-AzApplicationGateway -Name $AppGWName -ResourceGroupName $resourceGroup
$updatedGW = Add-AzApplicationGatewayCustomError -ApplicationGateway $AppGw -StatusCode HttpStatus502 -CustomErrorPageUrl $customError502Url
Set-AzApplicationGateway -ApplicationGateway $AppGw
